node/src/main/scala/coop/rchain/node/web/Transaction.scala

```diff
@@ -0,0 +1,235 @@
+import scodec.Codec
+import scodec.codecs.utf8
+
+import scala.collection.concurrent.TrieMap
+import scala.language.higherKinds
+
+final case class Transaction(
+    fromAddr: String,
+    toAddr: String,
+    amount: Long,
+    retUnforgeable: Par,
+    failReason: Option[String]
+)
+
+sealed trait TransactionType
+final case class PreCharge(deployId: String)       extends TransactionType
+final case class UserDeploy(deployId: String)      extends TransactionType
+final case class Refund(deployId: String)          extends TransactionType
+sealed trait SystemTransaction                     extends TransactionType
+final case class CloseBlock(blockHash: String)     extends SystemTransaction
+final case class SlashingDeploy(blockHash: String) extends SystemTransaction
+
+final case class TransactionInfo(transaction: Transaction, transactionType: TransactionType)
+
+final case class TransactionResponse(data: Seq[TransactionInfo])
+
+trait TransactionAPI[F[_]] {
+  def getTransaction(blockHash: String): F[Seq[TransactionInfo]]
+}
+
+/**
+  * This API is totally based on how RevVault.rho is written. If the `RevVault.rho` is re-written or changed,
+  * this API might end up with useless.
+  */
+final case class TransactionAPIImpl[F[_]: Sync: Concurrent](
+    blockReportAPI: BlockReportAPI[F],
+    // The transferUnforgeable can be retrieved based on the deployer and the timestamp of RevVault.rho
+    // in the genesis ceremony.
+    transferUnforgeable: Par
+) extends TransactionAPI[F] {
+  private def findUserDeployTransaction(
+      d: DeployInfoWithEventData,
+      report: SingleReport,
+      index: Int
+  ) = {
+    val transactions = findTransactions(report)
+    index match {
+      // the index order is based on how we play a user deploy
+      case 0 =>
+        transactions.map(t => TransactionInfo(t, PreCharge(d.deployInfo.get.sig))).pure
+      case 1 =>
+        transactions.map(t => TransactionInfo(t, UserDeploy(d.deployInfo.get.sig))).pure
+      case 2 =>
+        transactions.map(t => TransactionInfo(t, Refund(d.deployInfo.get.sig))).pure
+      case _ =>
+        Sync[F].raiseError[Vector[TransactionInfo]](
+          new Exception(
+            "It is not possible that user deploy got more than 3 reports(precharge, user, refund)."
+          )
+        )
+    }
+  }
+
+  def getTransaction(blockHash: String): F[Seq[TransactionInfo]] =
+    for {
+      report <- blockReportAPI.blockReport(blockHash, forceReplay = false)
+      ts <- report match {
+             case Right(b) =>
+               for {
+                 udts <- b.deploys.toList.traverse(d => {
+                          d.report.zipWithIndex.toList.traverse {
+                            case (report, index) => findUserDeployTransaction(d, report, index)
+                          }
+                        })
+                 userTransactions = udts.flatten.flatten
+                 systemDeployTransactions = b.systemDeploys.flatMap(
```

Comment:
I found it very handy to use curly brackets here so ` s =>` is in the same line as `b.systemDeploys.flatMap`, this way you save one + line space inside map body.